File: frontend/src/lib/api/edinetdb-config.ts
import { getJSON, postJSON, putJSON } from "./_client";

// EDINET DB（edinetdb.jp）接続設定（api_key/plan を DB+WebUI で管理・ADR-064・docs/api.md
// 「EDINET DB 設定」）。型は backend Pydantic（routers/edinetdb_config.py）と 1:1。公式 EDINET
// （env の edinet_api_key）とは別系統。#2 売掛/在庫の質の財務取得に使う。api_key は GET で常にマスク。

/** 契約プラン名（services/edinetdb_config._PLAN_LIMITS と 1:1・ドロップダウンと一致）。 */
export type EdinetdbPlan = "free" | "pro";

/** EDINET DB 接続の現在値（api_key はマスク済み・ADR-064）。 */
export type EdinetdbConfig = {
  api_key_masked: string; // "…AB12"（末尾4桁）・空鍵は ""
  has_api_key: boolean;
  plan: string; // free/pro
  configured: boolean; // api_key があり #2 取得が動くか
};

/** EDINET DB 疎通テストのレスポンス（POST /diagnostics/edinetdb-test・ADR-064）。 */
export type EdinetdbTestResponse = {
  configured: boolean; // API キーが設定されているか（false なら未設定）
  ok: boolean; // 認証が通り会社一覧が取れたか（configured=false のときは常に false）
  detail: string; // 人間向けメッセージ（成功＝収載社数＋月残予算／失敗＝エラー要旨）
};

/** EDINET DB 接続の現在値（api_key はマスク済み）。 */
export function getEdinetdbConfig(signal?: AbortSignal): Promise<EdinetdbConfig> {
  return getJSON<EdinetdbConfig>("/edinetdb/config", signal);
}

/** EDINET DB 接続を更新（api_key は空送信で据え置き＝write-only）。 */
export function updateEdinetdbConfig(body: {
  api_key?: string;
  plan?: EdinetdbPlan;
}): Promise<EdinetdbConfig> {
  return putJSON<EdinetdbConfig>("/edinetdb/config", body);
}

/** edinetdb.jp に認証ピングを 1 発投げる（DB 非依存・ADR-011「複数の起動口」）。
 * configured=false は未設定、ok=false は疎通失敗。detail を呼び出し側で表示する。 */
export function sendEdinetdbTest(): Promise<EdinetdbTestResponse> {
  return postJSON<EdinetdbTestResponse>("/diagnostics/edinetdb-test", {});
}

File: backend/app/routers/edinetdb_config.py
"""EDINET DB（edinetdb.jp）接続設定の REST ルータ（ADR-064・backend-router-pattern）。

設計の真実: docs/decisions.md ADR-064・docs/api.md「EDINET DB 設定」節。

`/settings` の WebUI から edinetdb.jp の api_key と契約プラン名（free/pro）を編集する。HTTP 入出力
だけの薄い層で、解決ロジックは services/edinetdb_config、クエリは db/repo/edinetdb_config が持つ
（ADR-005/014）。秘密の api_key は GET では必ずマスクし、更新は write-only（空送信は据え置き＝
jquants_config と同方針）。疎通テストは POST /diagnostics/edinetdb-test を流用する（このルータには
持たない）。公式 EDINET（DB の edinet_config・/edinet/config・ADR-087）とは別系統。
"""

from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy import Connection

from app.db import repo
from app.db.engine import get_conn, get_engine

router = APIRouter(tags=["edinetdb-config"])

# 許容プラン名（ADR-064・services/edinetdb_config._PLAN_LIMITS と 1:1）。UI のドロップダウンと一致。
EdinetDbPlan = Literal["free", "pro"]


class EdinetDbConfigOut(BaseModel):
    """EDINET DB 接続の公開表現（api_key はマスク・ADR-064）。"""

    api_key_masked: str  # "…AB12"（末尾 4 桁）。空鍵は ""
    has_api_key: bool
    plan: str  # free/pro
    configured: bool  # api_key があり #2 取得が動くか


class EdinetDbConfigUpdate(BaseModel):
    api_key: str | None = None  # None/空文字＝据え置き（write-only・ADR-064）
    plan: EdinetDbPlan | None = None


def _mask(api_key: str) -> str:
    """api_key をマスクする（GET で生キーを返さない・ADR-064・jquants_config と同方針）。"""
    if not api_key:
        return ""
    if len(api_key) <= 4:
        return "•" * len(api_key)
    return "…" + api_key[-4:]


def _config_out(conn: Connection) -> EdinetDbConfigOut:
    """edinetdb_config の現在値を表示用にまとめる（api_key はマスク・ADR-064）。"""
    row = repo.get_edinetdb_config(conn) or {}
    key = str(row.get("api_key") or "")
    plan = (str(row.get("plan") or "").strip().lower()) or "free"
    return EdinetDbConfigOut(
        api_key_masked=_mask(key),
        has_api_key=bool(key),
        plan=plan,
        configured=bool(key),
    )


@router.get("/edinetdb/config", response_model=EdinetDbConfigOut)
def get_edinetdb_config(conn: Connection = Depends(get_conn)) -> EdinetDbConfigOut:
    """EDINET DB 接続の現在値を返す（api_key はマスク・ADR-064）。"""
    return _config_out(conn)


@router.put("/edinetdb/config", response_model=EdinetDbConfigOut)
def update_edinetdb_config(body: EdinetDbConfigUpdate) -> EdinetDbConfigOut:
    """EDINET DB 接続を部分更新する（api_key は write-only＝空送信は据え置き・ADR-064）。"""
    with get_engine().begin() as conn:
        fields: dict[str, object] = {}
        if body.plan is not None:
            fields["plan"] = body.plan
        if body.api_key:  # 非空文字列のときだけ更新（空・None は据え置き＝write-only）
            fields["api_key"] = body.api_key
        if fields:
            repo.upsert_edinetdb_config(conn, fields)
        out = _config_out(conn)
    return out
